pkg/kubectl/cmd/get.go

```diff
@@ -48,66 +51,136 @@ Examples:
+	if isWatch || isWatchOnly {
+		r := resource.NewBuilder(f.Mapper, f.Typer, ClientMapperForCommand(cmd, f)).
+			NamespaceParam(GetKubeNamespace(cmd)).DefaultNamespace().
+			SelectorParam(selector).
+			ResourceTypeOrNameArgs(args...).
+			SingleResourceType().
+			Do()
+
+		mapping, err := r.ResourceMapping()
+		checkErr(err)
+
+		printer, err := printerForMapping(f, cmd, mapping)
+		checkErr(err)
+
+		obj, err := r.Object()
+		checkErr(err)
+
+		// print the current object
+		if !isWatchOnly {
+			if err := printer.PrintObj(obj, out); err != nil {
+				checkErr(fmt.Errorf("unable to output the provided object: %v", err))
+			}
+		}
 
-			client, err := f.RESTClient(cmd, mapping)
+		// print watched changes
+		if isWatch || isWatchOnly {
+			rv, err := mapping.MetadataAccessor.ResourceVersion(obj)
 			checkErr(err)
 
-			outputFormat := GetFlagString(cmd, "output")
-			templateFile := GetFlagString(cmd, "template")
-			defaultPrinter, err := f.Printer(cmd, mapping, GetFlagBool(cmd, "no-headers"))
+			w, err := r.Watch(rv)
 			checkErr(err)
 
-			outputVersion := GetFlagString(cmd, "output-version")
-			if len(outputVersion) == 0 {
-				outputVersion = mapping.APIVersion
-			}
+			kubectl.WatchLoop(w, func(e watch.Event) error {
+				return printer.PrintObj(e.Object, out)
+			})
+		}
+		return
+	}
 
-			printer, err := kubectl.GetPrinter(outputFormat, templateFile, outputVersion, mapping.ObjectConvertor, defaultPrinter)
-			checkErr(err)
+	printer, generic, err := printerForCommand(cmd)
+	checkErr(err)
 
-			restHelper := resource.NewHelper(client, mapping)
-			var obj runtime.Object
-			if len(name) == 0 {
-				obj, err = restHelper.List(namespace, labelSelector)
-			} else {
-				obj, err = restHelper.Get(namespace, name)
-			}
-			checkErr(err)
+	b := resource.NewBuilder(f.Mapper, f.Typer, ClientMapperForCommand(cmd, f)).
+		NamespaceParam(GetKubeNamespace(cmd)).DefaultNamespace().
+		SelectorParam(selector).
+		ResourceTypeOrNameArgs(args...).
+		Latest()
 
-			isWatch, isWatchOnly := GetFlagBool(cmd, "watch"), GetFlagBool(cmd, "watch-only")
+	if generic {
+		// the outermost object will be converted to the output-version
+		printer = kubectl.NewVersionedPrinter(printer, api.Scheme, outputVersion(cmd))
 
-			// print the current object
-			if !isWatchOnly {
-				if err := printer.PrintObj(obj, out); err != nil {
-					checkErr(fmt.Errorf("unable to output the provided object: %v", err))
-				}
-			}
+		obj, err := b.Flatten().Do().Object()
+		checkErr(err)
 
-			// print watched changes
-			if isWatch || isWatchOnly {
-				rv, err := mapping.MetadataAccessor.ResourceVersion(obj)
-				checkErr(err)
+		err = printer.PrintObj(obj, out)
+		checkErr(err)
+		return
+	}
 
-				w, err := restHelper.Watch(namespace, rv, labelSelector, labels.Everything())
-				checkErr(err)
+	// use the default printer for each object
+	err = b.Do().Visit(func(r *resource.Info) error {
```

Comment:
FYI: I'd eventually like it to be possible to invoke all of the business logic of kubectl via library call, not just through the command line. That will facilitate programmatic use of Kubernetes, such as to manage dataflow pipelines, CI pipelines, background worker pools, sophisticated rollouts, etc.

Command-line processing obviously has to be coupled to cobra and the kubectl command. Input/output should be less coupled. All the actual processing and client calls should be completely independently callable.

This particular batch of PRs don't need to be changed, but please keep this in mind as we evolve kubectl functionality.